virtual_rainforest/core/model.py

```diff
@@ -0,0 +1,116 @@
+"""The `core.model` module.
+
+The `core.model` module defines the api that all individual models (e.g. the soil
+model) should conform to. This consists of a class (`Model`), which defines the expected
+functions. The relevant modules will create classes to represent specific models, which
+will inherit from the `Model` base class. These subclasses will generally overwrite the
+functions defined in the base class, which are defined mainly to force a consistent api
+between models. It also establishes a model registry that allows models to become
+accessible across scripts without individual loading in.
+"""
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Any, Callable
+
+from numpy import datetime64, timedelta64
+
+from virtual_rainforest.core.logger import LOGGER
+
+MODEL_REGISTRY: dict[str, Callable] = {}
+"""A registry for different models."""
+
+
+class BaseModel(ABC):
+    """A superclass for all `vr` models.
+
+    Describes the common functions and attributes that all `vr` models should have. This
+    includes functions to setup, spin up and solve the specific model, as well as a
+    function to cleanup redundant model data. At this level these functions are not
+    define and are mere placeholders to be overwritten (where appropriate) by the
+    inheriting classes.
+
+    Args:
+        start_time: Point in time that the model simulation should be started.
+        end_time: Time that the model simulation should end
+        update_interval: Time to wait between updates of the model state.
+
+    Attributes:
+        name: Names the model that is described
+    """
+
+    name = "base"
+    # TODO - Once higher level timing function is written use it to set this
+    last_update = datetime64("2000-01-01")
+
+    def __init__(self, *args: Any, **kwargs: Any):
+        # Save args and kwargs, so that they can be found by the __repr__
+        self._args = args
+        self._kwargs = kwargs
+
+        self.update_interval: timedelta64 = args[0]
```

Comment:
This is very fragile - what if the first positional argument is not a date. Or if there are no positional arguments at all! 

Why did you remove the explicit arguments you had before? I think that was a better approach. Explicit is always better than implicit. And it also duplicates the storage of information, as the inputs are include din the explicit attributes and in args/kwargs. Just saving sometime when creating the `repr` string does not justify that, in my opinion. 

I'd suggest the following approach, in combination with the comments below: include an attribute storing the attribute strings that need to be included in `repr`, and use that.

```suggestion
    def __init__(self, update_interval: timedelta64, **kwargs: Any):
        self.update_interval = update_interval
        self._repr = ["update_interval"]
```